### ISSUE TITLE
Checking dependencies

### ISSUE THREAD

**Issue by olexandr-konovalov** [gap-packages/nq#8]
A user reported to @grahamknockillaree (trying to install HAP, for which nq is a dependency), that after

    ./configure

worked fine, `make` gave the following error:
```
 CCLD     nq
/usr/bin/ld: cannot find -lz
/usr/bin/ld: cannot find -lreadline
collect2: error: ld returned 1 exit status
make: *** [Makefile:537: nq] Error 1
```
On ubuntu 20.04. This was fixed by installing some additional ubuntu packages. Are there any undocumented dependencies needed to compile nq?

**Comment by fingolfin** [comment]
Neither library is used by nq. I assume those linker flags are somehow pulled in from GAP. Can't say more right now (typing this on my phone)

**Comment by fingolfin** [comment]
So yeah, those FLAGS are copied from GAP. So what I don't understand is how they were able to compile GAP. Perhaps the GAP installation was copied from another machine?

**Comment by junkicide** [comment]
I was the one who had reported this issue. I used apt to install gap.

**Comment by fingolfin** [comment]
OK, thanks! So the problem is that those FLAGS in GAP's sysinfo.gap reference files which are not installed now... This is a flaw which needs to be addressed in GAP (I started work on that for other reasons in https://github.com/gap-system/gap/pull/4044; it will have to be further refined).

In the meantime, I guess I could work around this by filtering out any `-lFOOBAR` flags in the GAP_CFLAGS which are different from `-lgmp`... THat's of course rather a hack, but it's by far the easiest and quickest fix I can think of right now.

**Comment by narc-Ontakac2** [comment]
The packages to install for compiling on debian are libreadline-dev, libzadc-dev. Just to provide an easy workaround until this is fixed.

**Comment by fingolfin** [comment]
Actually it is libreadline-devel and zlib1g-dev

**Comment by narc-Ontakac2** [comment]
My debuild tells me that all GAP_LIBS except -lgmp are not used:

dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/gap-nq/usr/bin/nq was not linked against libreadline.so.7 (it uses none of the library's symbols)
dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/gap-nq/usr/bin/nq was not linked against libm.so.6 (it uses none of the library's symbols)
dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/gap-nq/usr/bin/nq was not linked against libz.so.1 (it uses none of the library's symbols)
dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/gap-nq/usr/bin/nq was not linked against libdl.so.2 (it uses none of the library's symbols)
dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/gap-nq/usr/bin/nq was not linked against libutil.so.1 (it uses none of the library's symbols)

So it might be an option to simply do 
nq_LDADD = 
and add -lgmp using LIBS. I'll send a patch if I find out how to do this.

**Comment by fingolfin** [comment]
Yeah, they are indeed not used. Don't bother about that patch, it won't be merged unless it also takes care of the case were GAP was built with its own GMP (which the simplistic solution you described does not). Of course for Debian purposes this foes not matter, as you'll alway build against a system installed GMP, so it makes sense for to use such a patch, but it's not a general solution.

**Comment by fingolfin** [comment]
(Just to be clear: I appreciate your offer to send a patch, just wanted to make sure you don't end up wasting your time!)

**Comment by narc-Ontakac2** [comment]
I have to do the patch anyway. Debuild does have a way to patch the upstream source and a package with unnecessary dependencies is not acceptable as a debian package. I am currently working on [my own RFP](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=968545). NQ is a dependency for HAP to me.

**Comment by fingolfin** [comment]
After coming a variant of thisacross this myself today, I think the solution I'd prefer (and will one day work on ... perhaps...) is to add a "regular" `--with-gmp=PREFIX` but *in addition* have `--with-gaproot=` (but optional); if `--with-gmp=PREFIX` is given and with a PREFIX, we use that; if it is not given / or just with `yes` as prefix, then we try the usual things to "find" GMP. And if that fails and `--with-gaproot` is given, we try to figure out if GAP was built with its own GMP, and try to use that. One can vary a bit with the order in which things are tried, but I'd like to make the following three scenarios work seamlessly:
1. `./configure && make` should work if GMP is installed in `/usr` or `/usr/local` (and possibly a few more search trees, e.g. `/opt/homebrew`
2. `./configure --with-gmp=PREFIX && make` should work if GMP is installed in PREFIX
3. `./configure --with-gaproot=GAPROOT && make` should work if GAP was built with its own GMP

Bonus points if input 3 also works if GAP was not built with its own GMP, but rather GAP just found GMP as in approach 1. Double bonus points if it even works when GAP was built as in option 2 (i.e.: figure out what was passed to GAP's `configure` via `--with-gmp`). However, that's probably either requiring a hack, or supporting changes in the GAP build system. In any case, it's purely optional.

**Comment by fingolfin** [comment]
ohh, I just realized I did something like this for `anupq` so perhaps I can just crib some of that code

**Comment by fingolfin** [comment]
Resolves by PR #17
